File: simp-api-ingredients/schemas/product.py
from pydantic import BaseModel
import uuid
from typing import List, Optional, Dict

class ProductCreate(BaseModel):
    """Schema for creating a new product."""
    retail_id: Optional[int] = None
    src_product_id: Optional[uuid.UUID] = None  # Matches DB field
    english_name: str
    spanish_name: str
    amount: float
    weight: float
    measurement: str
    company_prices: Dict[str, float] = {}  # Mapping of company names to prices

class CompanyOut(BaseModel):
    """Schema for returning company details linked to a product."""
    company_id: uuid.UUID
    company_name: str
    price: float

class ProductOut(BaseModel):
    """Schema for returning product details."""
    product_id: uuid.UUID
    retail_id: Optional[int] = None
    src_product_id: Optional[uuid.UUID] = None  # Matches DB field
    english_name: str
    spanish_name: str
    amount: float
    weight: float
    measurement: str
    companies: List[CompanyOut] = []  # List of associated companies with prices

    class Config:
        from_attributes = True  # Ensures SQLAlchemy -> Pydantic conversion

class ProductCompanyOut(BaseModel):
    """Schema for returning linked product-company relationships."""
    product_id: uuid.UUID  # Matches DB field
    company_id: uuid.UUID
    company_name: str
    price: float

File: simp-front-end/src/lib/types/product.ts
export type Product = {
  product_id: string; // UUID
  retail_id?: number | null; // Optional
  src_product_id?: string | null; // UUID, Optional
  english_name: string;
  spanish_name: string;
  amount: number; // Default 1
  weight: number;
  measurement: string;
  companies?: ProductCompanyDetail[]; // Optional list of linked companies with names
};

export type ProductCompany = {
  product_id: string; // UUID (ForeignKey)
  company_id: string; // UUID (ForeignKey)
  price: number;
};

export type ProductCompanyDetail = {
  company_id: string;
  company_name: string;
  price: number;
};

export type ProductCreate = {
  retail_id?: number | null;
  src_product_id?: string | null; // Optional
  english_name: string;
  spanish_name: string;
  amount: number;
  weight: number;
  measurement: string;
  company_prices: Record<string, number>; // { "Company A": 10.5, "Company B": 12.0 }
};
